File: 200. Number of Islands.cpp
class Solution {
private:
    int row;
    int col;
public:
    // // DFS: O(MN)time & space, 22.03%, ok
    // int numIslands(vector<vector<char>>& grid) {
    //     if(!grid.size() || !grid[0].size()) return 0;
    //     row = grid.size(), col = grid[0].size();
    //     int ret = 0;
    //     vector<vector<bool>> visited(row, vector<bool>(col, 0));
    //     for(int i = 0; i < row; i++){
    //         for(int j = 0; j < col; j++){
    //             if(!visited[i][j] && grid[i][j] == '1'){
    //                 ret++;
    //                 dfs(grid, visited, i, j);
    //             }
    //             else visited[i][j] = 1;
    //         }
    //     }
    //     return ret;
    // }

    // void dfs(vector<vector<char>>& grid, vector<vector<bool>>& visited, int i, int j){
    //     if(i < 0 || i >= row || j < 0 || j >= col || visited[i][j] ) return;
    //     visited[i][j] = 1;
    //     if(grid[i][j] == '1'){
    //         dfs(grid, visited, i - 1, j);
    //         dfs(grid, visited, i + 1, j);
    //         dfs(grid, visited, i, j - 1);
    //         dfs(grid, visited, i, j + 1);
    //     }
    // }

    // DFS: reuse the original matrix for flagging: O(MN)time, O(1)space, 74.16%, ok
    int numIslands(vector<vector<char>>& grid) {
        if(!grid.size() || !grid[0].size()) return 0;
        row = grid.size(), col = grid[0].size();
        int ret = 0;
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if(grid[i][j] == '1'){
                    ret++;
                    dfs(grid, i, j);
                }
            }
        }
        return ret;
    }

    void dfs(vector<vector<char>>& grid, int i, int j){
        if(grid[i][j] == '0') return;
        grid[i][j] = '0';
        if(i > 0) dfs(grid, i - 1, j);
        if(j > 0) dfs(grid, i, j - 1);
        if(i < row - 1) dfs(grid, i + 1, j);
        if(j < col - 1) dfs(grid, i, j + 1);
    }

    // Union Find: to be continued

};

File: 200. Number of Islands.py
class Solution(object):
    
    # dfs with flag: O(mn)time & space
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        cnt, visited = 0, [[0 for j in xrange(len(grid[0]))] for i in xrange(len(grid))]
        def dfs(i, j):
            if 0 <= i < len(grid) and 0 <= j < len(grid[0]) and not visited[i][j] and grid[i][j] == '1':
                visited[i][j] = 1
                dir = [1, 0, -1, 0, 1]
                map(dfs, map(lambda x: x+i, dir[:-1]), map(lambda x: x+j, dir[1:]))
            return None
            
        for i in xrange(len(grid)):
            for j in xrange(len(grid[0])):
                if not visited[i][j] and grid[i][j] == '1':
                    dfs(i, j)
                    cnt += 1
        return cnt

    # dfs: O(mn) time, O(1) space
    def numIslands(self, grid):
        cnt = 0
        def dfs(i, j):
            if 0 <= i < len(grid) and 0 <= j < len(grid[0]) and grid[i][j] == '1':
                grid[i][j] = '0'
                dir = [1, 0, -1, 0, 1]
                map(dfs, map(lambda x: x+i, dir[:-1]), map(lambda x: x+j, dir[1:]))
            return None
            
        for i in xrange(len(grid)):
            for j in xrange(len(grid[0])):
                if grid[i][j] == '1':
                    dfs(i, j)
                    cnt += 1
        return cnt  
    
    # union find: O(MN) time, O(1) space
    